pkg/volume/util/operationexecutor/operation_generator_test.go

```diff
@@ -0,0 +1,250 @@
+		},
+	}
+
+	for _, tc := range testcases {
+		expectedPluginName := tc.pluginName
+		if tc.isCsiMigrationEnabled {
+			defer featuregatetesting.SetFeatureGateDuringTest(t, utilfeature.DefaultFeatureGate, features.CSIMigration, true)()
+			defer featuregatetesting.SetFeatureGateDuringTest(t, utilfeature.DefaultFeatureGate, tc.csiMigrationFeature, true)()
+			expectedPluginName = fmt.Sprintf("%s:%s", csi.CSIPluginName, tc.csiDriverName)
+		}
+		volumePluginMgr, plugin, tmpDir := initTestPlugins(t, tc.probVolumePlugins, tc.pluginName)
+		defer os.RemoveAll(tmpDir)
+
+		operationGenerator := getTestOperationGenerator(volumePluginMgr)
+
+		pod := &v1.Pod{ObjectMeta: metav1.ObjectMeta{UID: types.UID(string(uuid.NewUUID()))}}
+		volumeToUnmount := getTestVolumeToUnmount(pod, tc.pvSpec, tc.pluginName)
+
+		if tc.isCsiMigrationEnabled {
+			// GenerateUnmapVolumeFunc call blockVolumePlugin.NewBlockVolumeUnmapper and when the plugin is csi,
+			// csi plugin looks a file that contains some information about the volume,
+			// and GenerateUnmapVolumeFuncfails if csi plugin can't find that file.
+			// So the reason for calling plugin.NewBlockVolumeMapper for csi enabled case is creating that file.
+			csiSpec, err := translateSpec(volumeToUnmount.VolumeSpec)
+			if err != nil {
+				t.Fatalf("Can't translate volume to CSI")
+			}
+			_, mapperError := (*plugin).(volume.BlockVolumePlugin).NewBlockVolumeMapper(csiSpec, pod, volume.VolumeOptions{})
+			if mapperError != nil {
+				t.Fatalf("mapper error: %v\n", mapperError)
+			}
+		}
+
+		unmapVolumeFunc, e := operationGenerator.GenerateUnmapVolumeFunc(volumeToUnmount, nil)
+		if e != nil {
+			t.Fatalf("Error occurred while generating unmapVolumeFunc: %v", e)
+		}
+
+		metricFamilyName := "storage_operation_status_count"
+		labelFilter := map[string]string{
+			"status":         "success",
+			"operation_name": "unmap_volume",
+			"volume_plugin":  expectedPluginName,
+		}
+		// compare the relative change of the metric because of the global state of the prometheus.DefaultGatherer.Gather()
+		storageOperationStatusCountMetricBefore := findMetricWithNameAndLabels(metricFamilyName, labelFilter)
+
+		var ee error
+		unmapVolumeFunc.CompleteFunc(&ee)
+
+		storageOperationStatusCountMetricAfter := findMetricWithNameAndLabels(metricFamilyName, labelFilter)
+		if storageOperationStatusCountMetricAfter == nil {
+			t.Fatalf("Couldn't find the metric with name(%s) and labels(%v)", metricFamilyName, labelFilter)
+		}
+
+		if storageOperationStatusCountMetricBefore == nil {
+			assert.Equal(t, float64(1), *storageOperationStatusCountMetricAfter.Counter.Value, tc.name)
+		} else {
+			metricValueDiff := *storageOperationStatusCountMetricAfter.Counter.Value - *storageOperationStatusCountMetricBefore.Counter.Value
+			assert.Equal(t, float64(1), metricValueDiff, tc.name)
+		}
+	}
+}
+
+func findMetricWithNameAndLabels(metricFamilyName string, labelFilter map[string]string) *io_prometheus_client.Metric {
```

Comment:
are there no helper functions that exist in the codebase that do this exact thing? It's fine if the answer is no, but I'm just curious. Want to avoid duplicating code as much as possible

Reply:
I found [some functions](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/admission/metrics/testutil_test.go) that do the similar things but couldn't be sure about refactoring them to reuse, wdyt?

Reply:
I think it's fine to not refactor. I was hoping there would be some common util functions that already exist but it seems like that's not the case.